array_api_tests/test_special_cases.py

```diff
@@ -0,0 +1,646 @@
+    ),
+)
+
+
+def parse_unary_docstring(docstring: str) -> List[UnaryCase]:
+    match = r_special_cases.search(docstring)
+    if match is None:
+        return []
+    lines = match.group(1).split("\n")[:-1]
+    cases = []
+    for line in lines:
+        if m := r_case.match(line):
+            case = m.group(1)
+        else:
+            warn(f"line not machine-readable: '{line}'")
+            continue
+        if m := r_unary_case.search(case):
+            try:
+                case = UnaryCase.from_strings(*m.groups())
+            except ValueParseError as e:
+                warn(f"not machine-readable: '{e.value}'")
+                continue
+            cases.append(case)
+        elif m := r_even_int_round_case.search(case):
+            cases.append(even_int_round_case)
+        else:
+            if not r_remaining_case.search(case):
+                warn(f"case not machine-readable: '{case}'")
+    return cases
+
+
+class BinaryCond(Protocol):
+    def __call__(self, i1: float, i2: float) -> bool:
+        ...
+
+
+class BinaryResultCheck(Protocol):
+    def __call__(self, i1: float, i2: float, result: float) -> bool:
+        ...
+
+
+@dataclass(repr=False)
+class BinaryCase(Case):
+    expr: str
+    cond: BinaryCond
+    check_result: BinaryResultCheck
+
+
+r_special_cases = re.compile(
+    r"\*\*Special [Cc]ases\*\*(?:\n.*)+"
+    r"For floating-point operands,\n+"
+    r"((?:\s*-\s*.*\n)+)"
+)
+r_case = re.compile(r"\s+-\s*(.*)\.\n?")
+r_binary_case = re.compile("If (.+), the result (.+)")
+r_remaining_case = re.compile("In the remaining cases.+")
+
+r_cond_sep = re.compile(r"(?<!``x1_i``),? and |(?<!i\.e\.), ")
+r_cond = re.compile("(.+) (?:is|have) (.+)")
+
+r_element = re.compile("x([12])_i")
+r_input = re.compile(rf"``{r_element.pattern}``")
+r_abs_input = re.compile(rf"``abs\({r_element.pattern}\)``")
+r_and_input = re.compile(f"{r_input.pattern} and {r_input.pattern}")
+r_or_input = re.compile(f"either {r_input.pattern} or {r_input.pattern}")
+
+r_result = re.compile(r"(?:is|has a) (.+)")
+
+r_input_is_array_element = re.compile(
+    f"{r_array_element.pattern} is {r_array_element.pattern}"
+)
+r_both_inputs_are_value = re.compile("are both (.+)")
+
+
+def parse_binary_case(case_m: Match) -> BinaryCase:
+    cond_strs = r_cond_sep.split(case_m.group(1))
+    conds = []
+    cond_exprs = []
+    for cond_str in cond_strs:
+        if m := r_input_is_array_element.match(cond_str):
+            in_sign, input_array, value_sign, value_array = m.groups()
+            assert in_sign == "" and value_array != input_array  # sanity check
+            expr = f"{in_sign}x{input_array}ᵢ == {value_sign}x{value_array}ᵢ"
+            if value_array == "1":
+                if value_sign != "-":
+
+                    def cond(i1: float, i2: float) -> bool:
+                        eq = make_eq(i1)
+                        return eq(i2)
+
+                else:
```

Comment:
Might be a good idea to code these more defensively like

```py
if value_sign == '+':
    ...
elif value_sign == '-':
    ...
else:
    raise ValueParseError
```

Reply:
Hmm, I believe the regex match should guarantee that value_sign is only "", "-" or "+". A `raise` here does work well as a sanity check tho.